include/boost/geometry/algorithms/detail/gc_group_elements.hpp

```diff
@@ -0,0 +1,196 @@
+                {
+                    gc_element_id id2 = {1, qit->second};
+                    adjacent[id1].insert(id2);
+                    adjacent[id2].insert(id1);
+                }
+            }, boost::begin(gc1_view) + i);
+        }
+    }
+    else // group elements from the same GCs and the other GC
+    {
+        auto const rtree1 = gc_make_rtree_indexes(gc1_view, strategy);
+        for (auto it1 = rtree1.begin() ; it1 != rtree1.end() ; ++it1)
+        {
+            auto const b1 = it1->first;
+            gc_element_id id1 = {0, it1->second};
+            for (auto qit2 = rtree2.qbegin(index::intersects(b1)); qit2 != rtree2.qend(); ++qit2)
+            {
+                gc_element_id id2 = {1, qit2->second};
+                adjacent[id1].insert(id2);
+                adjacent[id2].insert(id1);
+            }
+            for (auto qit1 = rtree1.qbegin(index::intersects(b1)); qit1 != rtree1.qend(); ++qit1)
+            {
+                if (id1.gc_id != qit1->second)
+                {
+                    gc_element_id id11 = {0, qit1->second};
+                    adjacent[id1].insert(id11);
+                    adjacent[id11].insert(id1);
+                }
+            }
+        }
+        for (auto it2 = rtree2.begin(); it2 != rtree2.end(); ++it2)
+        {
+            auto const b2 = it2->first;
+            gc_element_id id2 = {1, it2->second};
+            for (auto qit2 = rtree2.qbegin(index::intersects(b2)); qit2 != rtree2.qend(); ++qit2)
+            {
+                if (id2.gc_id != qit2->second)
+                {
+                    gc_element_id id22 = {1, qit2->second};
+                    adjacent[id2].insert(id22);
+                    adjacent[id22].insert(id2);
+                }
+            }
+        }
+    }
+
+    // Traverse the graph and build connected groups i.e. groups of intersecting envelopes
+    std::deque<gc_element_id> queue;
+    std::array<std::vector<bool>, 2> visited = {
+        std::vector<bool>(boost::size(gc1_view), false),
+        std::vector<bool>(boost::size(gc2_view), false)
+    };        
+    for (auto const& elem : adjacent)
+    {
+        std::vector<gc_element_id> group;
+        if (! visited[elem.first.source_id][elem.first.gc_id])
+        {
+            queue.push_back(elem.first);
+            visited[elem.first.source_id][elem.first.gc_id] = true;
+            group.push_back(elem.first);
+            while (! queue.empty())
+            {
+                gc_element_id e = queue.front();
+                queue.pop_front();
+                for (auto const& n : adjacent[e])
+                {
+                    if (! visited[n.source_id][n.gc_id])
+                    {
+                        queue.push_back(n);
+                        visited[n.source_id][n.gc_id] = true;
+                        group.push_back(n);
+                    }
+                }
+            }
+        }
+        if (! group.empty())
```

Comment:
or
```
if (! group.empty() && ! intersecting_fun(group) ) return;
```

Reply:
Yes, this would be shorter but IDK really that it'd be better. The original makes it more clear that this is a call of a function passed by the user and not another condition check. It is also more similar to the second call of the other function below. @barendgehrels what's your opinion?